### PR TITLE
Log `UnknownBlock` errors from runtime

### PR DESCRIPTION
**Pull request by tdimitrov**
This PR is building upon https://github.com/paritytech/substrate/pull/12707 and is related to https://github.com/paritytech/polkadot/issues/5885

# The problem

All runtime calls first check the runtime version before issuing the actual runtime api call. If a call is made for a block which is pruned (e.g. because a better fork was finalized), the runtime api version can't be checked and `NotSupported` error is returned. This is misleading because the problem is different. The call IS supported but it can't be executed at a pruned block.

# How this is solved

This PR adds explicit logging for `UnknownBlock` runtime errors so that it is clear why the runtime call is failing. Furthermore `UnknownBlock` variant is added to `RuntimeApiError` enum in `node/subsystem-types/src/errors.rs`. This allows any subsystem to handle such errors in its own way.

# What this PR doesn't do

The PR doesn't fix the root cause (prevent executing logic on pruned blocks). 

# Sample logs

E.g. for `{node=~"versi-validator-f.*"} |~ "UnknownBlock|discarded"`:

image

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2244337

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2315557

**Comment by ordian** [comment]
What is the rationale for using `warns` for such errors?

> The PR doesn't fix the root cause (prevent executing logic on pruned blocks).

How is it possible to do that in general case? Wouldn't you have a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) problem?

**Review by ordian — APPROVED** [review]
I misunderstood the point of the issue was to make the error more explicit and not silence it. However, I am still not sure what would the point of warning the user about this if this error is expected to happen as we already observe in the logs.

File: node/core/runtime-api/src/lib.rs [hunk 1]
```diff
@@ -355,33 +355,55 @@ where
...
+							0
+						}
+					}
 				});
 
-			let res = if runtime_version >= version {
-				client.$api_name(relay_parent $(, $param.clone() )*).await
+			let res = match runtime_version {
+				Ok(runtime_version) if runtime_version >= version => {
+					client.$api_name(relay_parent $(, $param.clone() )*).await
```
Comment (BradleyOlson64) [inline node/core/runtime-api/src/lib.rs:393]: Mind explaining what's going on here? I'm not familiar with the syntax being used. Particularly the $ and * uses.

**Review by BradleyOlson64 — COMMENTED** [review]
Looks clean!

**Review by BradleyOlson64 — APPROVED** [review]

File: node/core/runtime-api/src/lib.rs (hunk 1 above)
Comment (tdimitrov) [reply node/core/runtime-api/src/lib.rs:393]: Yes.... the code is part of a macro. Check the declaration [here](https://github.com/paritytech/polkadot/pull/6288/files#diff-1c816593e067f095a3388e13824a0729ac4d1d9b7fcf96ab2ea6761ba423a9dcR353)

It's a huge topic, but in nutshell - this is a declarative macro. Think about it as macros in C but  on heavy steroids. The syntax is close to pattern matching - you can call the macro with different parameters. The variables with $ are the matched macro parameters. The parameters are [here](https://github.com/paritytech/polkadot/pull/6288/files#diff-1c816593e067f095a3388e13824a0729ac4d1d9b7fcf96ab2ea6761ba423a9dcR354).

`($($param:expr),*)` on line 354 means 'multiple expressions separated by comma' while `$(, $param.clone() )*` means 'put all the expressions in $param here and separate them with comma'. 

Really quick and lame explanation but it's a huge topic.

**Comment by tdimitrov** [comment]
These logs doesn't contribute much to the understandably of the logs. I'm closing the PR.

**Pull request closed without merging**